### PR TITLE
Improve test coverage for scheduler component.

### PR DESCRIPTION
**Pull request by svkcemk**

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java [hunk 1]
```diff
@@ -28,12 +28,43 @@
...
+        });
+    }
+
+    @Test
+    public void testDelayWithFixed() throws Exception {
+        await().atMost(4, TimeUnit.SECONDS).until(() -> {
+            String body = RestAssured.get("/scheduler/get-count").then().statusCode(200).extract().body().asString();
+            return Integer.parseInt(body) > 1;
+        });
+    }
```
Comment (ppalaga) [inline integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java:52]: It looks like you assume `testDelay()` to be executed before `testDelayWithFixed()` ? Maybe one test method would serve better?

```suggestion
    @Test
    public void testDelay() throws Exception {
        await().atMost(2, TimeUnit.SECONDS).until(() -> {
            String body = RestAssured.get("/scheduler/get-count").then().statusCode(200).extract().body().asString();
            return Integer.parseInt(body) > 1;
        });
        await().atMost(2, TimeUnit.SECONDS).until(() -> {
            String body = RestAssured.get("/scheduler/get-count").then().statusCode(200).extract().body().asString();
            return Integer.parseInt(body) > 2;
        });
    }
```

File: integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/ShoppingList.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    public void addItems() {
+        int item = new Random().nextInt(100);
+        lists.add(item);
+    }
+
+    public void removeItems() {
+        if (lists != null && lists.size() > 0) {
+            lists.remove(lists.size() - 1);
+        }
+    }
```
Comment (ppalaga) [inline integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/ShoppingList.java:38]: This does not seem to be used. So we may remove it, no?

File: integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/ShoppingList.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    public List<Integer> getItems() {
+        return lists;
+    }
```
Comment (ppalaga) [inline integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/ShoppingList.java:42]: There is a potential race between `getItems()` and `addItems()` - you may get concurrent modification exception when accessing the list via `getItems()`. You may want to switch from ArrayList to java.util.concurrent.CopyOnWriteArrayList.

File: integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/ShoppingList.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Named;
+
+@ApplicationScoped
+@Named("shoppingList")
+public class ShoppingList {
+    List<Integer> lists = new ArrayList<>();
+
+    public void addItems() {
```
Comment (ppalaga) [inline integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/ShoppingList.java:29]: You are adding just a single item.
```suggestion
    public void addItem() {
```

File: integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/GreedyCounter.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+public class GreedyCounter {
+    private int count = 0;
+
+    public int getCount() {
+        return count;
+    }
+
+    public void increment() {
+        count++;
+    }
```
Comment (ppalaga) [inline integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/GreedyCounter.java:33]: Note that getCount and increment() can be called from differen threads concurrently. While I do not think it matters much in how we call them from tests currently, it is still potentially dangerous (end users may copy this and use differently, we may extend the test in the future). So producing a named AtomicInteger is perhaps a better choice.

File: integration-test-groups/foundation/scheduler/pom.xml [hunk 6]
```diff
@@ -34,6 +34,10 @@
         <dependency>
             <groupId>org.apache.camel.quarkus</groupId>
             <artifactId>camel-quarkus-scheduler</artifactId>
+        </dependency>
+            <dependency>
+            <groupId>org.apache.camel.quarkus</groupId>
+            <artifactId>camel-quarkus-bean</artifactId>
         </dependency>
```
Comment (ppalaga) [inline integration-test-groups/foundation/scheduler/pom.xml:41]: I'd vote for doing without this additional dependency. We could e.g. have a an ad hoc processor to increment the counter, something like:

```
.process(e -> myCounter.incrementAndGet());
```

File: integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerRoute.java [hunk 7]
```diff
@@ -30,11 +31,24 @@ public class SchedulerRoute extends RouteBuilder {
     @Inject
     @Named("schedulerCounter")
     AtomicInteger schedulerCounter;
+    @Inject
+    CamelContext context;
```
Comment (ppalaga) [inline integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerRoute.java:35]: The context field does not seem to be used. Can we perhaps remove it?

File: integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerRoute.java [hunk 8]
```diff
@@ -30,11 +31,24 @@ public class SchedulerRoute extends RouteBuilder {
...
     public void configure() throws Exception {
-        from("scheduler:start?initialDelay=1")
+        from("scheduler:withInitialDelay?initialDelay=1")
                 .process(e -> schedulerCounter.incrementAndGet());
 
+        from("scheduler:withDelay?delay=1000")
+                .to("bean:myCounter?method=increment");
+
+        from("scheduler:useFixedDelay?initialDelay=2000&useFixedDelay=true")
+                .to("bean:myCounter?method=increment");
```
Comment (ppalaga) [inline integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerRoute.java:46]: I'd vote for using shorter delays, to save some testing time. 200ms should be fine. At the same time, we should not check using `==` in the tests, only `<`, for the case that the test is too slow to observe the exact value.

**Review by ppalaga — COMMENTED** [review]
Thanks @svkcemk , some comments inline.

**Comment by aldettinger** [comment]
The Camel quarkus documentation is talking about [forcing scheduler to be idle](https://camel.apache.org/components/3.18.x/scheduler-component.html#_forcing_the_scheduler_to_be_idle).

Would it make sense to test this one ? Maybe that's another PR.

File: integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/GreedyCounter.java (hunk 5 above)
Comment (aldettinger) [reply integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/GreedyCounter.java:33]: And maybe the proposal is even to produce one `AtomicInteger` named `greedyCounter` and a second one named `mycounter`

**Comment by aldettinger** [comment]
That's great start @svkcemk , keep up the good work.
Also note that CI issues might not be related to this PR.

**Comment by svkcemk** [comment]
Thanks @ppalaga @aldettinger for review, sorry for late update ( I was on PTO last week ) updated the PR based o the review. @aldettinger I am creating a separate PR for the case : Forcing the scheduler to be idle.

**Comment by ppalaga** [comment]
Sorry, no time to review, because I am on a conference. If @aldettinger finds this good, I find it good either :)

File: integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java [hunk 9]
```diff
@@ -28,12 +28,42 @@
...
         await().atMost(5, TimeUnit.SECONDS).until(() -> {
             String body = RestAssured.get("/scheduler/get").then().statusCode(200).extract().body().asString();
             return !body.equals("0");
         });
     }
 
+    @Test
+    public void testDelay() throws Exception {
+
+        await().atMost(2, TimeUnit.SECONDS).until(() -> {
```
Comment (aldettinger) [inline integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java:41]: Could the body could be > 1 even if withDelay is never called ?
At the end of the day, we could maybe better to separate `withDelay` and `useFixedDelay`.
So, 2 distinct tests and 2 distinct atomic integers.
What do you think ?

File: integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java [hunk 10]
```diff
@@ -28,12 +28,42 @@
...
+            String body = RestAssured.get("/scheduler/get-count").then().statusCode(200).extract().body().asString();
+            return Integer.parseInt(body) > 2;
+        });
+    }
+
+    @Test
+    public void testDelayWithRepeat() throws Exception {
+        await().atMost(4, TimeUnit.SECONDS).until(() -> {
+            String body = RestAssured.get("/scheduler/get-items").then().statusCode(200).extract().body().asString();
+            return body.split(",").length >= 4;
```
Comment (aldettinger) [inline integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java:56]: Here, we expect a list of items >= 4. Can't we do that with a simple AtomicInteger ?

File: integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java (hunk 9 above)
Comment (svkcemk) [reply integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java:41]: @aldettinger do you mean two separate Atomic int like myCounter or its just during testing two separate int one for withDelay and one for withFixedDelay ?

Comment (aldettinger) [reply integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java:41]: The main idea is that, from my point of view, the current implementation of this test could pass even when the `withDelay` route is never called.
From there, one solution could be to have two distinct AtomicInteger, let's say `withDelayCounter` and `useFixedDelayCounter`. And then, it's possible to check that each route has been invoked separately.

File: integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerResource.java [hunk 11]
```diff
@@ -33,11 +31,15 @@
...
+    AtomicInteger withDelayCounter;
 
-    @Named("shoppingList")
-    CopyOnWriteArrayList<Integer> shoppingList;
+    @Inject
+    @Named("useFixedDelayCounter")
+    AtomicInteger useFixedDelayCounter;
+
+    @Named("withDelayRepeat")
+    AtomicInteger withDelayRepeat;
```
Comment (aldettinger) [inline integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerResource.java:42]: Maybe `withDelayRepeatCounter` would align with other names ? What do you think ?

File: integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java [hunk 12]
```diff
@@ -28,12 +28,43 @@
...
 
+    @Test
+    public void testDelay() throws Exception {
+        await().atMost(2, TimeUnit.SECONDS).until(() -> {
+            String body = RestAssured.get("/scheduler/get-delay-count").then().statusCode(200).extract().body()
+                    .asString();
+            return Integer.parseInt(body) > 2;
+        });
+
+        await().atMost(2, TimeUnit.SECONDS).until(() -> {
```
Comment (aldettinger) [inline integration-test-groups/foundation/scheduler/src/test/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerTest.java:46]: And so, at the end of the day. Those 2 tests are independent ? So we could better separate them back in tests ? Like `testDelay` and `testFixedDelay` ?

File: integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerResource.java (hunk 11 above)
Comment (svkcemk) [reply integration-test-groups/foundation/scheduler/src/main/java/org/apache/camel/quarkus/component/scheduler/it/SchedulerResource.java]: Yes sounds good!

**Review by aldettinger — APPROVED** [review]
Cool, it's already rebased to a single commit.
Let's check the build and we can merge then.
Well done @svkcemk :clap:

**Comment by aldettinger** [comment]
The failing test is not related. Merging then.

**Pull request merged**